feat: Scene event metric tests in tools integration test project

josiemessa opened the pull request.

Commit:
Create initial Unity project

Commit:
Remove half-written test

Commit:
Add scene event metric tests

Commit:
Actually commit tests

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
remove duplicate asmdef file

Commit:
Scene unload test now passing, also check in project settings as scene must be in build settings for tests to work

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
Unload scene and first scene sync tests pass

Commit:
More scene sync tests

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
Clean up tests to work with new scene sync metric structure

Commit:
Add teardown to unload loaded scenes between tests

josiemessa commented:
Current state of this PR:

* Test written for all scene event types other than `S2C_Resync` - need to look how to trigger this but hopefully will look very similar to `S2C_Sync`
* Have copied utility classes over from NGO runtime tests, but we may be able to just expose these using an update to AssemblyInfo to get around the current issues with using them where they are.
* Have added a helper filter to WaitForMetricValues so you can inspect the metric and filter based on values inside it, so this would also need to be copied into NGO runtime test version if we remove the custom version in the testproject-tools-integration folder.
* Need to move "while/yield returns" into static helper functions - there are other versions of these in various tests around but nothing as stripped down as this one.

### Problems 

The tests all pass when run in isolation, but if you run them together they fail. I've tracked down the issue as far as I can but don't have a good solution for it yet.

Initially I hit the issue that the scenes are still loaded from the previous test, meaning that the code blows up when we try to reload it as it's already seen it. I've added a teardown that should get around this problem, but this then causes problems in the setup of the next test as it tries to destroy any network objects with global hash ID 0. I have no idea how it's getting into this state, so will need someone to look into.

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
Tests now all pass, and using existing WaitForMetricValues

Commit:
Add resync test
Currently not passing as need to either spawn network object or load a scene
but struggling to sync up prefab lists/scene lists

Commit:
Try cloning game object to trigger resync

Commit:
despawn the network object as that is what the resync requires

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
Split scene event tests

Commit:
Add WaitForSceneEvent helper class

josiemessa commented on testproject-tools-integration/Assets/Tests/Runtime/SceneEventTests.cs:
```diff
@@ -95,44 +99,46 @@ public IEnumerator TestSceneEventMetrics_LoadScene_S2C_Load()
-                    clientSceneLoaded = true;
-            };
+            var waitForClientLoadComplete = new WaitForSceneEvent(
+                m_ServerNetworkSceneManager,
```
This should be m_ClientNetworkSceneManager

Commit:
Add helper for server load and unload scene

josiemessa commented on testproject-tools-integration/Assets/Tests/Runtime/SceneEventTests.cs:
```diff
@@ -95,44 +99,46 @@ public IEnumerator TestSceneEventMetrics_LoadScene_S2C_Load()
+        public IEnumerator TestC2SLoadCompleteReceived()
+        {
+            var waitForClientLoadComplete = new WaitForSceneEvent(
+                m_ServerNetworkSceneManager,
```
Should be m_ClientNetworkSceneManager

josiemessa commented on testproject-tools-integration/Assets/Tests/Runtime/SceneEventTests.cs:
```diff
@@ -216,58 +204,74 @@ public IEnumerator TestSceneEventMetrics_LoadScene_S2C_LoadComplete()
             m_ServerNetworkSceneManager.OnSceneEvent += sceneEvent =>
             {
-                if (sceneEvent.SceneEventType != SceneEventData.SceneEventTypes.S2C_Unload) return;
+                if (sceneEvent.SceneEventType != SceneEventData.SceneEventTypes.S2C_Unload)
```
I don't know why I wrote this one differently to the others but we could make it look like the same callback in S2C_Load:
```
            m_ServerNetworkSceneManager.OnSceneEvent += sceneEvent =>
            {
                if (sceneEvent.SceneEventType.Equals(SceneEventData.SceneEventTypes.S2C_Load))
                {
                    serverSceneLoaded = sceneEvent.AsyncOperation.isDone;
                    sceneEvent.AsyncOperation.completed += _ => serverSceneLoaded = true;
                }
            };
```

Also means we could make this another helper class like WaitForSceneEvent

josiemessa commented on testproject-tools-integration/Assets/Tests/Runtime/SceneEventTests.cs:
```diff
@@ -430,28 +414,37 @@ public IEnumerator TestSceneEventMetrics_SyncScenes_S2C_Sync()
+        [UnityTest]
+        public IEnumerator TestS2CSyncReceived()
+        {
+            MultiInstanceHelpers.CreateNewClients(1, out var newClients);
```
Looking at these again I think we might want to teardown the new clients between tests as there's a possibility they could interfere with other tests as they'll still be around, they aren't at the moment though so maybe it's safe to keep them?

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

NoelStephensUnity reviewed (commented):
Would like to see some comments on what each test is supposed to do, just so anyone else that might need to come back to the test(s) can ramp up faster.

josiemessa commented on com.unity.netcode.gameobjects/Tests/Runtime/Metrics/Utility/WaitForMetricValues.cs:
```diff
@@ -1,11 +1,9 @@
-#if MULTIPLAYER_TOOLS
```
Unintentionally removed this and it's now blowing up the build so will fix!

Commit:
Merge remote-tracking branch 'origin/feature/MTT-748-scene-tests' into feature/MTT-748-scene-tests

Commit:
Add code comments to help readability, address PR comments and fix issue with WaitForMetricValues.cs where #if directive had been unintentionally removed.

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
Fix internalsvisibleto

Commit:
formatting

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
Use latest tools package

Commit:
Merge branch 'feature/MTT-748-scene-tests' of github.com:Unity-Technologies/com.unity.multiplayer.mlapi into feature/MTT-748-scene-tests

becksebenius-unity commented on com.unity.netcode.gameobjects/Tests/Runtime/Metrics/Utility/WaitForMetricValues.cs:
```diff
@@ -17,13 +17,24 @@ internal class WaitForMetricValues<TMetric> : IMetricObserver
+
+        Filter m_FilterDelegate;
+
+
```
Super nit, lots of extra whitespace here

becksebenius-unity commented on testproject-tools-integration/Assets/Tests/Runtime/SceneEventTests.cs:
```diff
@@ -0,0 +1,704 @@
+        [UnityTearDown]
+        public override IEnumerator Teardown()
+        {
+
```
Super nit, extra whitespace

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
Use updated version of tools package
Also: remove network manager from scene, as network managers are created by multi-instance test framework so do not need to be in the scene

JesseOlmer commented on com.unity.netcode.gameobjects/Tests/Runtime/AssemblyInfo.cs:
```diff
@@ -0,0 +1,5 @@
+using System.Runtime.CompilerServices;
```
Why are we using internalsvisibleto? Shouldn't we just make those things public? I mean, it's a test assembly...

becksebenius-unity replied:
I could be wrong about why Josie did it this way but Package Verification was throwing warnings about undocumented public APIs for test assemblies in the tools package, so we made them all internal and used exposed internals for cross-assembly references to address that.

josiemessa replied:
The classes in the test assembly make use of internal classes from the tools/NGO repos, so when I made the test classes public there was then a mismatch between the levels of accessibility. So this seemed the best way around the issue.

JesseOlmer commented on testproject-tools-integration/Assets/Tests/Runtime/testproject.toolsintegration.runtimetests.asmdef:
```diff
@@ -0,0 +1,23 @@
+        {
+            "name": "com.unity.multiplayer.tools",
+            "expression": "",
+            "define": "MULTIPLAYER_TOOLS"
```
Under what circumstances is this expected to be false in this project? Or is it just a "shortcut" to define it for non-tools tests?

becksebenius-unity replied:
I'm guessing this is just following suit on how we enable this define in other areas. Its a good point though, this define is probably not necessary for this assembly - we could just write the code unconditionally since the whole premise of this project is that tools is installed.

josiemessa replied:
I can try defining without the conditional and see if that behaves itself, I had a bit of a nightmare trying to get this to compile and tried many different permutations to get this to work so this may have been unnecessarily left in!

josiemessa replied:
Have removed this and it seems to be working OK

JesseOlmer commented on testproject-tools-integration/ProjectSettings/ProjectVersion.txt:
```diff
@@ -1,2 +1,2 @@
-m_EditorVersion: 2021.2.0b4
-m_EditorVersionWithRevision: 2021.2.0b4 (af9ec38f7da3)
+m_EditorVersion: 2021.2.0b12
```
Is running on the beta intentional for this project?

becksebenius-unity replied:
2021.2 is still in beta I think? I believe the latest is 2021.2.0b13 though.

josiemessa replied:
This has been picked up from the version I was using, so which version should this be referencing? I think the test project is on 2020.3? That said I'm currently having issues getting the profiler to run on 2020.3!

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

Commit:
Remove unnecessary version define and let unity generate asmdef

Commit:
Wait for sync to complete before verifying metrics to avoid race condition affecting client IDs

Commit:
fix comment

Commit:
Merge branch 'develop' into feature/MTT-748-scene-tests

JesseOlmer commented on testproject-tools-integration/Assets/Tests/Runtime/testproject.toolsintegration.runtimetests.asmdef:
```diff
@@ -0,0 +1,25 @@
+        "Unity.Multiplayer.MetricTypes",
+        "Unity.Multiplayer.NetStats",
+        "UnityEngine.TestRunner",
+        "UnityEditor.TestRunner"
```
nit: your runtime tests probably shouldn't depend on things from UnityEditor, right? Not exactly sure the context of "runtime" tests for tools tests. If this dll is expected to work in a standalone build, it won't with this dependency.

josiemessa replied:
It looks like when you use unity to edit the asmdef, for some reason it brings all of this in. So no, it wasn't intentionally added. I haven't tried running this standalone before. Do you want me to remove the dependency now or check that in later?

JesseOlmer replied:
We can deal with it later. Can you make sure there's a tracking issue filed somewhere so we don't forget about it? This ended up biting us pretty badly when we enabled platform CI for mobile & consoles.

josiemessa replied:
https://jira.unity3d.com/browse/MTT-1383 done!

JesseOlmer reviewed (approved):
lgtm
